### PR TITLE
add a bunch of tests with array typed columns to CalciteArraysQueryTest

### PR DESCRIPTION
**Pull request by clintropolis**
### Description

Duplicates a bunch of tests in `CalciteArraysQueryTest` to test real array columns using some of the same data used by the limited array tests of `CalciteNestedDataQueryTest` to get a lot more coverage on array typed columns.

In the process discovered a bug with filter pushdown in `UnnestStorageAdapter` when filtering on unnested array columns, where it was pushing down the filter on the unnested output column when the input was an array typed column, which is totally incorrect. I think i've modified things to behave correctly...

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/test/java/org/apache/druid/segment/UnnestStorageAdapterTest.java [hunk 1]
```diff
@@ -516,6 +516,30 @@ public void test_not_pushdown_not_filter()
...
+                and(ImmutableList.of(
+                    selector("newcol", "3"),
+                    selector(COLUMNNAME, "7")
+                ))
+            )),
+            selector(OUTPUT_COLUMN_NAME, "1")
+        ))
+    ));
+    testComputeBaseAndPostUnnestFilters(
+        testQueryFilter,
```
Comment (soumyava) [inline processing/src/test/java/org/apache/druid/segment/UnnestStorageAdapterTest.java:537]: in a sense pushdown is still happening here, albeit not the entire thing. Would have been good if we can have a filter on an expression involving more than one column to show that the pushdown in basically nil. But that can be a bit complicated

**Review by soumyava — COMMENTED** [review]
Thanks for raising this PR. I'll recommend adding a test with unnest over a JOIN involving arrays and then a JOIN over an unnest, just for the sake of completeness.

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java [hunk 2]
```diff
@@ -327,8 +327,7 @@ to generate filters to be passed to base cursor (filtersPushedDownToBaseCursor)
               (BooleanFilter) queryFilter,
               inputColumn,
               inputColumnCapabilites,
-              filterSplitter,
-              isTopLevelAndFilter
+              filterSplitter
```
Comment (soumyava) [inline processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:330]: We should remove all occurrences of `isTopLevelAndFilter` which is not doing anything

File: processing/src/test/java/org/apache/druid/segment/UnnestStorageAdapterTest.java (hunk 1 above)
Comment (clintropolis) [reply processing/src/test/java/org/apache/druid/segment/UnnestStorageAdapterTest.java:537]: oops, i started to write a test by copying a test and then didn't change anything, my bad, will fix

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java [hunk 3]
```diff
@@ -322,19 +322,17 @@ to generate filters to be passed to base cursor (filtersPushedDownToBaseCursor)
...
           );
           // If rewite on entire query filter is successful then add entire filter to preFilter else skip and only add to post filter.
-          if (filterSplitter.getPreFilterCount() == filterSplitter.getOriginalFilterCount()) {
+          if (!preFilterList.isEmpty()) {
             if (queryFilter instanceof AndFilter) {
               filterSplitter.addPreFilter(new AndFilter(preFilterList));
-            } else if (queryFilter instanceof OrFilter) {
+            } else if (queryFilter instanceof OrFilter && filterSplitter.getPreFilterCount() == filterSplitter.getOriginalFilterCount()) {
```
Comment (soumyava) [inline processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:335]: Nit. we can put a comment with an example here like if query is like
```
select * from table, UNNEST(MV_TO_ARRAY(dim3)) as u(d3) where m1=1 OR d3='a'
```
The pre filter after rewrite will be like `m1=1 OR dim3='a'` and the count of clauses in the original and pre filter matching means a successful rewrite. We can also highlight how is this for AND

**Review by soumyava — APPROVED** [review]
I think intelliJ inspection is complaining about asList() on a single element, can be replaced by  Collections.singletonList(). Rest LGTM ! Thanks for the tests this uncovered the filter pushdown logic issue

**Comment by soumyava** [comment]
Also would be good to add a join on top of unnest using array typed columns.

**Comment by pranavbhole** [comment]
looks good to me

**Pull request merged**